crates/solver/src/liquidity/zeroex.rs

```diff
@@ -9,39 +9,87 @@ use {
         liquidity_collector::LiquidityCollecting,
         settlement::SettlementEncoder,
     },
-    anyhow::Result,
+    anyhow::{anyhow, Context, Result},
     contracts::{GPv2Settlement, IZeroEx},
+    ethcontract::Address,
+    futures::{SinkExt, StreamExt},
     model::{order::OrderKind, TokenPair},
     primitive_types::{H160, U256},
+    reqwest::Url,
     shared::{
         ethrpc::Web3,
         http_solver::model::TokenAmount,
         recent_block_cache::Block,
-        zeroex_api::{Order, OrderRecord, OrdersQuery, ZeroExApi},
+        zeroex_api::{
+            websocket::{OrderRecord, OrderState, OrdersResponse},
+            Order,
+            OrdersQuery,
+            ZeroExApi,
+        },
     },
     std::{
         collections::{HashMap, HashSet},
         sync::Arc,
+        time::Duration,
     },
+    tokio::{net::TcpStream, sync::Mutex},
+    tokio_tungstenite::{tungstenite::Message, MaybeTlsStream, WebSocketStream},
+    tracing::Instrument,
 };
 
 pub struct ZeroExLiquidity {
-    pub api: Arc<dyn ZeroExApi>,
     pub zeroex: IZeroEx,
     pub gpv2: GPv2Settlement,
     pub allowance_manager: Box<dyn AllowanceManaging>,
+    // cached orders, updated by background task
+    pub cache: Arc<Mutex<HashMap<OrderId, OrderRecord>>>,
 }
 
+// hash generated by the 0x api
+type OrderId = String;
 type OrderBuckets = HashMap<(H160, H160), Vec<OrderRecord>>;
+const DEFAULT_BACKOFF: Duration = Duration::from_secs(3);
 
 impl ZeroExLiquidity {
     pub fn new(web3: Web3, api: Arc<dyn ZeroExApi>, zeroex: IZeroEx, gpv2: GPv2Settlement) -> Self {
         let allowance_manager = AllowanceManager::new(web3, gpv2.address());
+        let cache: Arc<Mutex<HashMap<OrderId, OrderRecord>>> = Default::default();
+        let inner = cache.clone();
+        let api = api.clone();
+        let sender = gpv2.address();
+
+        tokio::task::spawn(
+            async move {
+                let mut backoff = DEFAULT_BACKOFF;
+                // loop needed for reconnections
+                loop {
+                    // initialize cache by fetching all existing orders via http api
+                    if let Err(err) = init_cache(api.clone(), inner.clone(), sender).await {
```

Comment:
Do we want to re-initialize the cache one each loop?

Reply:
Yes because if websocket fails reconnecting websocket only does not guarantee that we continue where we stopped, meaning we might have inconsistent state.